### DISCUSSION TITLE
How to launch .appinstaller using Appinstaller.exe from command line?

### DISCUSSION THREAD

**Discussion by HppZ** [microsoft/winget-cli#1591] [Q&A] [answered]
> "C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_1.16.12576.0_x64__8wekyb3d8bbwe\AppInstaller.exe" -ServerName:App.AppX9rwyqtrq9gw3wnmrap9a412nsc7145qh.mca

I got this from Task Manager.

**Comment by denelon** [comment, marked as answer]
Can you provide a bit more clarity on what you are/were trying to do?

**Reply by HppZ** [reply]
some special Windows 10 does not support double click to install appinstaller even has DesktopAppInstaller installed manually (Windows LTSC version does not have DesktopAppInstaller). Is file association broken? so I want to build a program to install DesktopAppInstaller and myapp.appinstaller for users.

**Reply by jedieaston** [reply]
You can do it via [Add-AppxPackage](https://docs.microsoft.com/en-us/powershell/module/appx/add-appxpackage?view=windowsserver2019-ps) in PowerShell. It's how we provision it in the sandbox. Just note you'll need to get some VCRedists and pass them as dependencies. 

[Here's one I use when the AppInstaller isn't around.](https://gist.github.com/jedieaston/0e64dd0ea27ba34d72e38538fee46c20)

**Reply by HppZ** [reply]
some Windows 10 need Developer Mode being enabled to install appinstaller, how to enable it in command line or code?

**Reply by jedieaston** [reply]
Here's an example in [PowerShell](https://stackoverflow.com/a/40033638).

**Reply by HppZ** [reply]
How to launch DesktopAppInstaller UI to install appinstaller? I see this from Task Manager:
>"C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_1.16.12576.0_x64__8wekyb3d8bbwe\AppInstaller.exe" -ServerName:App.AppX9rwyqtrq9gw3wnmrap9a412nsc7145qh.mca

It does not work if I run it directly and I don't know what is "ServerName:App.AppX9rwyqtrq9gw3wnmrap9a412nsc7145qh.mca"

**Reply by jvintzel** [reply]
The Desktop app installer should install on LTSC if you provide the framework packages and offline license file.  Right now we do not have a command line way to invoke the UX, the recommendation is PowerShell, as folks suggested above.

**Reply by HppZ** [reply]
I can run 'Add-AppxPackage -AppInstallerFile' without DesktopAppInstaller installed, so is it necessary to install DesktopAppInstaller?
What's the relation between ‘Add-AppxPackage’ command and DesktopAppInstaller？

**Reply by HppZ** [reply]
One of my user failed to run 'Add-AppxPackage -AppInstallerFile' and the error code is: 0x80D03002.

**Reply by jedieaston** [reply]
DesktopAppInstaller is a UI for interacting with the UWP infrastructure.

Is there anything notable in the log? `Get-AppxLog` should bring it up, or else check Event Viewer.

**Reply by HppZ** [reply]
Get-AppxLog output is (it is in Chinese):

Time                      ID           Message                                                                                                                                             
----                      --           -------                                                                                                                                             
2021/10/19 18:16:49       603          已使用主参数 bilibili_win.appinstaller 以及选项 0 和 0 开始对程序包执行部署 AddFromAppInstallerOperation 操作。要获取有关诊断应用部署问题的帮助，请参阅 http://go.microsoft.com/fwlink/?LinkId=2
                                       35160。                                                                                                                                              
2021/10/19 18:16:49       651          Appinstaller 操作失败，错误代码为 0x80D03002。详细信息: Unknown error                                                                                              
2021/10/19 18:16:49       404          程序包 file:///D:/Documents/Tencent%20Files/3179470669/FileRecv/bilibili_win.appinstaller 的 AppX 部署操作失败，错误为 0x80D03002。此失败的具体错误文本是: Appinstaller 操作失败，错
                                       误代码为 0x80D03002。详细信息: Unknown error

**Reply by jedieaston** [reply]
Do you know what build of Windows they are on? `$PSVersionTable.BuildVersion` should have it in PowerShell, or just use `winver`.

**Reply by HppZ** [reply]
I will contact that user and tell you later.

**Reply by HppZ** [reply]
版本	Windows 10 专业版
版本号	20H2
安装日期	‎2021/‎8/‎24
操作系统内部版本	19042.985
体验	Windows Feature Experience Pack 120.2212.2020.0

**Reply by HppZ** [reply]
@jedieaston

**Reply by HppZ** [reply]
I think the only way to solve that LTSC installation issue is to build my own installer EXE, any other option?

**Reply by jedieaston** [reply]
I've installed it on LTSC via that script before, still not sure what's up with this. Let me get a fresh image and try it again.

**Reply by HppZ** [reply]
another error: 
>please check whether the msixvc support services are installed

what is this?

**Reply by jedieaston** [reply]
I'm not sure why you'd be getting that error, but in other news, there's now instructions from MS on how to install winget for an entire machine, which also work if you're on LTSC: https://github.com/microsoft/winget-cli/blob/2b9f2f349155c4e0ad151e2b0a1d1fa3eb9c0019/doc/troubleshooting/README.md#machine-wide-provisioning

They worked for me, maybe try those ones and see if it clears it up. You're on the same LTSC build where you're getting this error, right?

**Reply by HppZ** [reply]
some of my users have problems with installation of appinstaller file, appinstaller is not reliable as win32 EXE, right?

**Reply by HppZ** [reply]
@jvintzel could you fix it?

**Reply by HppZ** [reply]
> The Desktop app installer should install on LTSC if you provide the framework packages and offline license file. Right now we do not have a command line way to invoke the UX, the recommendation is PowerShell, as folks suggested above.

using Powershell to install appinstaller sometimes does not work either.

**Reply by HppZ** [reply]
> another error:
> 
> > please check whether the msixvc support services are installed
> 
> what is this?

at least tell me which services AppInstaller depends to work?

**Reply by jedieaston** [reply]
That error implies there's something wrong with the appx deployment stuff on your machine (see this [long thread](https://techcommunity.microsoft.com/t5/msix-deployment/windows-10-2004-msix-not-updating-please-check-whether-the/m-p/1466701)). You don't have any updates pending, do you? 

Can you post the output of  `Get-AppxLog` from PowerShell?

**Reply by HppZ** [reply]
some of my users encountered this issue, not me.

**Reply by HppZ** [reply]
after manually installed DesktopAppInstaller, one of my user installed .appinstaller successfully. 

So is DesktopAppInstaller necessary?
DesktopAppInstaller VS Add-AppxPackage command?

**Reply by HppZ** [reply]
new error from my user (double click appinstaller file):
![image](https://user-images.githubusercontent.com/12285582/139397972-6e626ac9-e509-4072-af86-2da023af639b.png)

update:
using add-appxpackage -appinstaller command succeed.

**Reply by jvintzel** [reply]
We should move this conversation to the Microsoft Tech Community were we look at MSIX, as this is not specific to the Windows Package manager.  

https://techcommunity.microsoft.com/t5/msix-deployment/bd-p/MSIX-Deployment

If you can continue to reproduce the error its best to file a feedback hub item so we can get better data.

**Reply by HppZ** [reply]
what should I do if there is no FeedbackHub installed on user's machine?
